### PR TITLE
[CoreML] Adapt to `MLMultiArray.dataPointer` deprecation

### PR DESCRIPTION
**Pull request by NickLucche**
### Description
This PR addresses https://github.com/microsoft/onnxruntime/issues/17652.
The deprecated `MLMultiArray.dataPointer` is replaced with `.getBytesWithHandler`, as suggested by the docs.
For now, I am only checking that the output `MLMultiArray` is contiguous, returning unsupported operation when that is not the case. 
I think this is already better than what we have right now, so we can block unsafe calls to `.dataPointer` (if any..).

I would be happy to implement the handling of the non-contiguous case (replacing `memcpy` for such cases) as suggested by @edgchen1, but I am not sure how to reproduce that case to add a corresponding unit-test.
Would we have to define a custom `MLCustomLayer` to get a non-contiguous output from a model..?

### Motivation and Context
Fix https://github.com/microsoft/onnxruntime/issues/17652.

### CODE REVIEW

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 1]
```diff
@@ -169,6 +169,14 @@ Status CreateInputFeatureProvider(const std::unordered_map<std::string, OnnxTens
   conversion_buffers_out = std::move(conversion_buffers);
   return Status::OK();
 }
+
+bool IsArrayContiguous(MLMultiArray *array) {
+    int batch_stride = [array.strides[0] intValue];
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:174]: can we use `int64_t` for the element count variables?

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 2]
```diff
@@ -169,6 +169,14 @@ Status CreateInputFeatureProvider(const std::unordered_map<std::string, OnnxTens
...
+    auto shape = array.shape;
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:175]: https://github.com/microsoft/onnxruntime/blob/5a623dca0118a1fc75419875d2af813161d097a0/docs/Coding_Conventions_and_Standards.md?plain=1#L97
```suggestion
    auto* shape = array.shape;
```

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 3]
```diff
@@ -324,7 +332,12 @@ - (Status)predict:(const std::unordered_map<std::string, OnnxTensorData>&)inputs
                                  ") do not match");
         }
 
-        const void* model_output_buffer = data.dataPointer;
+        ORT_RETURN_IF_NOT(IsArrayContiguous(data),
+                            "Non contiguous output MLMultiArray are not currently supported");
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:336]:
```suggestion
                            "Non-contiguous output MLMultiArray is not currently supported");
```

File: onnxruntime/core/providers/coreml/model/model.mm
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:1]: please run clang-format on modified files.
one way is via lintrunner:
https://github.com/microsoft/onnxruntime/blob/main/docs/Coding_Conventions_and_Standards.md#linting

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 4]
```diff
@@ -324,7 +332,12 @@ - (Status)predict:(const std::unordered_map<std::string, OnnxTensorData>&)inputs
...
+        __block const void* model_output_buffer=nil;
+        [data getBytesWithHandler:^(const void *bytes, NSInteger size) {
+            model_output_buffer = bytes;
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:339]: should we also check that `size` is what we expect?

**Review by edgchen1 — COMMENTED** [review]
thanks for the PR! had a few comments.

**Comment by edgchen1** [comment]
> I would be happy to implement the handling of the non-contiguous case (replacing memcpy for such cases) as suggested by @edgchen1, but I am not sure how to reproduce that case to add a corresponding unit-test.
> Would we have to define a custom MLCustomLayer to get a non-contiguous output from a model..?

I think it's fine to only handle the contiguous case for now. Not clear to me how to test that either.

File: onnxruntime/core/providers/coreml/model/model.mm (hunk 4 above)
Comment (NickLucche) [reply onnxruntime/core/providers/coreml/model/model.mm:339]: Sure it won't hurt to be extra-sure and check for that too. Added it, it looks a bit ugly though as I have to check within the switch statement

**Comment by NickLucche** [comment]
Thanks a lot for reviewing the PR!

**Comment by NickLucche** [comment]
@microsoft-github-policy-service agree

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 5]
```diff
@@ -334,11 +349,13 @@ - (Status)predict:(const std::unordered_map<std::string, OnnxTensorData>&)inputs
         switch (onnx_data_type) {
           case ONNX_NAMESPACE::TensorProto_DataType_FLOAT: {
             const auto output_data_byte_size = num_elements * sizeof(float);
+            ORT_RETURN_IF_NOT(coreml_buffer_size == output_data_byte_size, "CoreML output buffer size and expected output size differ");
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:352]: nit: 120 character line length limit

https://github.com/microsoft/onnxruntime/blob/ac4e72604605be524f70d1e92da81b70c5db984a/docs/Coding_Conventions_and_Standards.md?plain=1#L7-L8

Comment (NickLucche) [reply onnxruntime/core/providers/coreml/model/model.mm:352]: sorry I didn't see this wasn't automatic on purpose
https://github.com/microsoft/onnxruntime/blob/ac4e72604605be524f70d1e92da81b70c5db984a/.clang-format#L6-L7

**Review by edgchen1 — DISMISSED** [review]

**Comment by edgchen1** [comment]
there are some errors from the MacOS CI build:

https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1156472&view=logs&j=07136875-e4b3-5a15-ff50-026d13f57d01&t=ff34eec9-4cff-535b-345b-da5da417f512

```
/Users/runner/work/1/s/onnxruntime/core/providers/coreml/model/model.mm:177:21: error: comparison of integers of different signs: 'int' and 'NSUInteger' (aka 'unsigned long') [-Werror,-Wsign-compare]
  for (int i = 1; i < [shape count]; i++) batch_elems *= [shape[i] longLongValue];
                  ~ ^ ~~~~~~~~~~~~~
/Users/runner/work/1/s/onnxruntime/core/providers/coreml/model/model.mm:339:15: error: 'getBytesWithHandler:' is only available on macOS 12.3 or newer [-Werror,-Wunguarded-availability-new]
        [data getBytesWithHandler:^(const void* bytes, NSInteger size) {
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Applications/Xcode_14.3.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/System/Library/Frameworks/CoreML.framework/Headers/MLMultiArray.h:125:1: note: 'getBytesWithHandler:' has been marked as being introduced in macOS 12.3 here, but the deployment target is macOS 10.14.0
- (void)getBytesWithHandler:(void (NS_NOESCAPE ^)(const void *bytes, NSInteger size))handler API_AVAILABLE(macos(12.3), ios(15.4), watchos(8.5), tvos(15.4)) NS_REFINED_FOR_SWIFT;
^
/Users/runner/work/1/s/onnxruntime/core/providers/coreml/model/model.mm:339:15: note: enclose 'getBytesWithHandler:' in an @available check to silence this warning
        [data getBytesWithHandler:^(const void* bytes, NSInteger size) {
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/runner/work/1/s/onnxruntime/core/providers/coreml/model/model.mm:352:50: error: comparison of integers of different signs: 'int64_t' (aka 'long long') and 'const unsigned long' [-Werror,-Wsign-compare]
            ORT_RETURN_IF_NOT(coreml_buffer_size == output_data_byte_size,
                              ~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~
/Users/runner/work/1/s/include/onnxruntime/core/common/common.h:202:19: note: expanded from macro 'ORT_RETURN_IF_NOT'
  ORT_RETURN_IF(!(condition), __VA_ARGS__)
                  ^~~~~~~~~
/Users/runner/work/1/s/include/onnxruntime/core/common/common.h:193:9: note: expanded from macro 'ORT_RETURN_IF'
    if (condition) {                                                                                           \
        ^~~~~~~~~
/Users/runner/work/1/s/onnxruntime/core/providers/coreml/model/model.mm:359:50: error: comparison of integers of different signs: 'int64_t' (aka 'long long') and 'const unsigned long' [-Werror,-Wsign-compare]
            ORT_RETURN_IF_NOT(coreml_buffer_size == output_data_byte_size,
                              ~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~
/Users/runner/work/1/s/include/onnxruntime/core/common/common.h:202:19: note: expanded from macro 'ORT_RETURN_IF_NOT'
  ORT_RETURN_IF(!(condition), __VA_ARGS__)
                  ^~~~~~~~~
/Users/runner/work/1/s/include/onnxruntime/core/common/common.h:193:9: note: expanded from macro 'ORT_RETURN_IF'
    if (condition) {                                                                                           \
        ^~~~~~~~~
/Users/runner/work/1/s/onnxruntime/core/providers/coreml/model/model.mm:370:50: error: comparison of integers of different signs: 'int64_t' (aka 'long long') and 'unsigned long' [-Werror,-Wsign-compare]
            ORT_RETURN_IF_NOT(coreml_buffer_size == num_elements * sizeof(int32_t),
                              ~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/runner/work/1/s/include/onnxruntime/core/common/common.h:202:19: note: expanded from macro 'ORT_RETURN_IF_NOT'
  ORT_RETURN_IF(!(condition), __VA_ARGS__)
                  ^~~~~~~~~
/Users/runner/work/1/s/include/onnxruntime/core/common/common.h:193:9: note: expanded from macro 'ORT_RETURN_IF'
    if (condition) {                                                                                           \
        ^~~~~~~~~
5 errors generated.
```

**Comment by NickLucche** [comment]
I run the ci_build pipeline on my side, looks ok. 
The only error to be addressed is this one:
```bash
'getBytesWithHandler:' is only available on macOS 12.3 or newer
note: 'getBytesWithHandler:' has been marked as being introduced in macOS 12.3 here, but the deployment target is macOS 10.14.0
```
What's your suggestion here? Do we check the target version and use the deprecated method for older builds or..? 

I dug a bit into the CI too but I am not sure if the macos version is left to azure availability to decide atm (couldn't spot a macos version, just `MacosArch`). Could be good to additionally build on latest macOS anyway (I take that back if it's already being done).

**Comment by edgchen1** [comment]
> I run the ci_build pipeline on my side, looks ok. The only error to be addressed is this one:
> 
> ```shell
> 'getBytesWithHandler:' is only available on macOS 12.3 or newer
> note: 'getBytesWithHandler:' has been marked as being introduced in macOS 12.3 here, but the deployment target is macOS 10.14.0
> ```
> 
> What's your suggestion here? Do we check the target version and use the deprecated method for older builds or..?
> 
> I dug a bit into the CI too but I am not sure if the macos version is left to azure availability to decide atm (couldn't spot a macos version, just `MacosArch`). Could be good to additionally build on latest macOS anyway (I take that back if it's already being done).

I think the deployment target is specified here:
https://github.com/microsoft/onnxruntime/blob/1bc115719c0d07ad658a9bb458ddf8f1bf379a41/tools/ci_build/github/azure-pipelines/templates/mac-cpu-packing-jobs.yml#L34

If we need to support older deployment targets, we can do something like you said, conditionally use the deprecated method.

We are building on a more recent MacOS version, but [MACOSX_DEPLOYMENT_TARGET](https://cmake.org/cmake/help/latest/envvar/MACOSX_DEPLOYMENT_TARGET.html) specifies the minimum version we support.

@snnn @pranavsharma @skottmckay What's the minimum MacOS version that we need to support?

**Comment by NickLucche** [comment]
Hey, any news on the best strategy here? :)

**Comment by edgchen1** [comment]
> Hey, any news on the best strategy here? :)

Sorry for the delay. Following up with the team about it.

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 6]
```diff
@@ -324,7 +332,14 @@ - (Status)predict:(const std::unordered_map<std::string, OnnxTensorData>&)inputs
...
         }
 
-        const void* model_output_buffer = data.dataPointer;
+        ORT_RETURN_IF_NOT(IsArrayContiguous(data),
+                          "Non-contiguous output MLMultiArray is not currently supported");
+        __block const void* model_output_buffer = nil;
+        __block unsigned long coreml_buffer_size = 0;
+        [data getBytesWithHandler:^(const void* bytes, NSInteger size) {
+          model_output_buffer = bytes;
+          coreml_buffer_size = size;
```
Comment (PetterS) [inline onnxruntime/core/providers/coreml/model/model.mm:341]: Hi! I believe you should use the `bytes` pointer while inside the handler and not pass it outside like this.

Comment (NickLucche) [reply onnxruntime/core/providers/coreml/model/model.mm:341]: Hi thanks for the feedback! I've been using it like this in the past as the docs don't specify any arc management nor lifetime for this pointer, so I assumed there were none. But I will fix it as you suggest as it seems safer in both cases!

**Comment by edgchen1** [comment]
Hi @NickLucche. It looks like ORT still needs to support a minimum version of MacOS 11 for now.

Could we check if the new API is available before using it (maybe with [`@available`](https://clang.llvm.org/docs/LanguageExtensions.html#objective-c-available))?

**Comment by NickLucche** [comment]
This looks like a clean solution, thanks for your help!

**Comment by NickLucche** [comment]
Just pushed a version using `@available` to check the api requirements at runtime so we can trigger CI and check that build runs through.
I can then implement @PetterS review, as that requires moving a bit more code around than anticipated due to variables capturing and error handling.

Minor note: If `skip_buffer_size_check` feels like too bloated we can remove it. I initially wanted to use `coreml_buffer_size` negative values to skip that check , but changing ulong to long required a cascading of static_casts I didn't like so much.

**Comment by edgchen1** [comment]
> Hey @edgchen1 , I am sorry but I can't quite reproduce the issue in the CI locally, I am trying to track this command
> 
> https://github.com/microsoft/onnxruntime/blob/3dece27f51f60f49f899a0321f019539168679ce/tools/ci_build/github/azure-pipelines/mac-ios-ci-pipeline.yml#L50
> 
> 
> I am on arm64 with xcode 14.3.1, but I couldn't gather much info on the 'macOS-13' CI agent which runs the jobs.
> Would you have any suggestions on how to best work with the CI or how to align local env with the CI one?

Sorry, didn't keep an eye on the build status. Not obvious to me why it failed, but the simulator tests have been flaky before so I restarted it. Let's see how that goes.

"macOS-13" is one of the Azure DevOps hosted agents.
https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml#software

**Comment by NickLucche** [comment]
Thanks for that, it appears last run had better luck in testing! ^^

I've went on and integrated the change suggested by @PetterS, basically I've refactored the MLMultiArray->OnnxTensor buffer copy into own function so we can keep both `getBytesWithHandler` and deprecate `dataPointer` and still make sense of what's going on.
Let me know what you think.

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 7]
```diff
@@ -23,6 +23,8 @@
 #include "core/providers/coreml/model/host_utils.h"
 #include "core/providers/coreml/shape_utils.h"
 
+#define HAS_GET_BYTES_WITH_HANDLER_API @available(macOS 12.3, iOS 15.4, *)
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:26]: do we need to introduce a macro if it's only used in one place?
maybe putting the `@available` inline and adding a comment about what it's guarding is good enough.

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 8]
```diff
@@ -169,6 +171,60 @@ Status CreateInputFeatureProvider(const std::unordered_map<std::string, OnnxTens
...
+
+bool IsArrayContiguous(MLMultiArray* array) {
+  int64_t batch_stride = [array.strides[0] longLongValue];
+  const auto* shape = array.shape;
+  int64_t batch_elems = 1;
+  for (unsigned long i = 1; i < shape.count; i++) batch_elems *= [shape[i] longLongValue];
+  return batch_stride == batch_elems;
+}
+
+onnxruntime::common::Status copyMLMultiArrayBuffer(const void* mlmultiarray_buffer, void* tensor_buffer,
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:183]:
```suggestion
Status CopyMLMultiArrayBuffer(const void* mlmultiarray_buffer, void* tensor_buffer,
```
- function naming convention is `MyFunctionName`
- no need to fully onnxruntime::common::Status here

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 9]
```diff
@@ -169,6 +171,60 @@ Status CreateInputFeatureProvider(const std::unordered_map<std::string, OnnxTens
...
+                                                   MLMultiArray* array_info, const OnnxTensorInfo& tensor_info,
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:184]:
```suggestion
                                                   const MLMultiArray* array_info, const OnnxTensorInfo& tensor_info,
```
if possible, can we make it a pointer to const? not sure if that works for objective-c types

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 10]
```diff
@@ -169,6 +171,60 @@ Status CreateInputFeatureProvider(const std::unordered_map<std::string, OnnxTens
...
+                                                   bool skip_buffer_size_check = true,
+                                                   const unsigned long mlmultiarray_buffer_size = 0) {
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:186]: can we use `std::optional<unsigned long>` instead? and avoid the default arguments if possible?

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 11]
```diff
@@ -169,6 +171,60 @@ Status CreateInputFeatureProvider(const std::unordered_map<std::string, OnnxTens
...
+      const auto output_span = gsl::span{static_cast<int64_t*>(tensor_buffer), num_elements};
+      std::transform(model_output_span.begin(), model_output_span.end(), output_span.begin(),
+                     [](int32_t v) { return static_cast<int64_t>(v); });
+      break;
+    }
+    default:
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
+                             "Output data type is not supported, actual type: ", onnx_data_type);
+  }
+  return onnxruntime::common::Status::OK();
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:226]:
```suggestion
  return Status::OK();
```

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 12]
```diff
@@ -298,23 +354,23 @@ - (Status)predict:(const std::unordered_map<std::string, OnnxTensorData>&)inputs
         return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "output_features has no value for ", output_name);
       }
 
-      auto* data = [output_value multiArrayValue];
+      __block MLMultiArray* data = [output_value multiArrayValue];
 
-      const auto coreml_static_output_shape = [&]() {
+      const auto coreml_static_output_shape = [](MLMultiArray* data) {
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:359]: just curious, do `__block` variables work differently for lambda captures?

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 13]
```diff
@@ -298,23 +354,23 @@ - (Status)predict:(const std::unordered_map<std::string, OnnxTensorData>&)inputs
         return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "output_features has no value for ", output_name);
       }
 
-      auto* data = [output_value multiArrayValue];
+      __block MLMultiArray* data = [output_value multiArrayValue];
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:357]: is `__block` only need for variables that we want to modify from the block?

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 14]
```diff
@@ -324,41 +380,20 @@ - (Status)predict:(const std::unordered_map<std::string, OnnxTensorData>&)inputs
...
-            std::transform(model_output_span.begin(), model_output_span.end(), output_span.begin(),
-                           [](int32_t v) { return static_cast<int64_t>(v); });
-            break;
-          }
-          default:
-            return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
-                                   "Output data type is not supported, actual type: ", onnx_data_type);
+        ORT_RETURN_IF_NOT(IsArrayContiguous(data),
+                          "Non-contiguous output MLMultiArray is not currently supported");
+        __block Status output_status;
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:385]: nit: can we rename this to `copy_status`?

**Review by edgchen1 — COMMENTED** [review]
nice refactoring. just had a few comments.

File: onnxruntime/core/providers/coreml/model/model.mm (hunk 10 above)
Comment (NickLucche) [reply onnxruntime/core/providers/coreml/model/model.mm:186]: absolutely!

File: onnxruntime/core/providers/coreml/model/model.mm (hunk 12 above)
Comment (NickLucche) [reply onnxruntime/core/providers/coreml/model/model.mm:359]: yes, variables with `__block` storage cannot be captured by & from c++ lambdas (compile error). I believe the `__block` identifier simply isn't understood on the cpp side, as the behavior shouldn't be different outside objc blocks interactions (there may be objc compiler optimizations that prove me wrong here though).

I should say we don't really need __block here though as var isn't mutable, realized it while adding the "const" I overlooked earlier, so thank you! :)

File: onnxruntime/core/providers/coreml/model/model.mm (hunk 13 above)
Comment (NickLucche) [reply onnxruntime/core/providers/coreml/model/model.mm:357]: yep only mutables sorry about that, see other conversation :)

**Comment by NickLucche** [comment]
Just pushed the changes, thanks a lot for reviewing you really caught some things I was about to overlook! Code should be better now than when we had previously :D  
Let me know if there's something else I could improve here

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 15]
```diff
@@ -169,6 +170,60 @@ Status CreateInputFeatureProvider(const std::unordered_map<std::string, OnnxTens
   conversion_buffers_out = std::move(conversion_buffers);
   return Status::OK();
 }
+
+bool IsArrayContiguous(MLMultiArray* array) {
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:174]:
```suggestion
bool IsArrayContiguous(const MLMultiArray* array) {
```

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 16]
```diff
@@ -298,9 +353,9 @@ - (Status)predict:(const std::unordered_map<std::string, OnnxTensorData>&)inputs
         return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "output_features has no value for ", output_name);
       }
 
-      auto* data = [output_value multiArrayValue];
+      MLMultiArray* data = [output_value multiArrayValue];
 
-      const auto coreml_static_output_shape = [&]() {
+      const auto coreml_static_output_shape = [&data]() {
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:358]: good change to be explicit in the capture list.
nit: if I see capture by reference, it's an indicator that the captured variable may be modified. as it's a pointer I think we can just capture by value.
```suggestion
      const auto coreml_static_output_shape = [data]() {
```

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 17]
```diff
@@ -313,8 +368,8 @@ - (Status)predict:(const std::unordered_map<std::string, OnnxTensorData>&)inputs
       const auto static_output_shape = GetStaticOutputShape(output_tensor_info.shape, coreml_static_output_shape,
                                                             *logger_);
 
-      void* output_buffer = get_output_tensor_mutable_raw_data_fn(output_name, output_tensor_info.data_type,
-                                                                  static_output_shape);
+      __block void* output_buffer = get_output_tensor_mutable_raw_data_fn(output_name, output_tensor_info.data_type,
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:371]: does this variable need to be declared with `__block`? the pointer itself is not modified in the block.

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 18]
```diff
@@ -10,6 +10,7 @@
 #include <cstdint>
 #include <unordered_map>
 #include <vector>
+#include <optional>
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:13]: nit: alphabetical order
```suggestion
#include <optional>
#include <unordered_map>
#include <vector>
```

Comment (NickLucche) [reply onnxruntime/core/providers/coreml/model/model.mm:13]: wish lintrunner did that..

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 19]
```diff
@@ -324,41 +379,21 @@ - (Status)predict:(const std::unordered_map<std::string, OnnxTensorData>&)inputs
...
-                           [](int32_t v) { return static_cast<int64_t>(v); });
-            break;
-          }
-          default:
-            return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
-                                   "Output data type is not supported, actual type: ", onnx_data_type);
+        ORT_RETURN_IF_NOT(IsArrayContiguous(data),
+                          "Non-contiguous output MLMultiArray is not currently supported");
+        __block Status copy_status;
+        const auto tensor_info = output_tensor_info;
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:385]: is this copy required for use in a block?

**Review by edgchen1 — APPROVED** [review]
looks good!

**Review by edgchen1 — DISMISSED** [review]

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 20]
```diff
@@ -324,41 +379,21 @@ - (Status)predict:(const std::unordered_map<std::string, OnnxTensorData>&)inputs
...
-                           [](int32_t v) { return static_cast<int64_t>(v); });
-            break;
-          }
-          default:
-            return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
-                                   "Output data type is not supported, actual type: ", onnx_data_type);
+        ORT_RETURN_IF_NOT(IsArrayContiguous(data),
+                          "Non-contiguous output MLMultiArray is not currently supported");
+        __block Status copy_status;
+        const auto *tensor_info = &output_tensor_info;
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/model/model.mm:385]:
```suggestion
        const auto* tensor_info = &output_tensor_info;
```

**Review by edgchen1 — APPROVED** [review]

**Comment by edgchen1** [comment]
@NickLucche thanks for your contribution and your patience with this PR. glad we got it merged!

**Comment by NickLucche** [comment]
Thank you for your help and patience! @edgchen1

**Pull request merged**